src/node_report_utils.cc

```diff
@@ -0,0 +1,576 @@
+              << " characters)\n";
+    return;
+  }
+  snprintf(report_filename, sizeof(report_filename), "%s", args);
+}
+
+/*******************************************************************************
+ * Function to process node-report config: specification of report directory.
+ ******************************************************************************/
+void ProcessNodeReportDirectory(const char* args) {
+  if (strlen(args) == 0) {
+    std::cerr << "Missing argument for node-report directory option\n";
+    return;
+  }
+  if (strlen(args) > NR_MAXPATH) {
+    std::cerr << "Supplied node-report directory path too long (max "
+              << NR_MAXPATH
+              << " characters)\n";
+    return;
+  }
+  snprintf(report_directory, sizeof(report_directory), "%s", args);
+}
+
+/*******************************************************************************
+ * Function to process node-report config: verbose mode switch.
+ ******************************************************************************/
+unsigned int ProcessNodeReportVerboseSwitch(const char* args) {
+  if (strlen(args) == 0) {
+    std::cerr << "Missing argument for node-report verbose switch option\n";
+    return 0;
+  }
+  // Parse the supplied switch
+  if (!strncmp(args, "yes", sizeof("yes") - 1) ||
+      !strncmp(args, "true", sizeof("true") - 1)) {
+    return 1;
+  } else if (!strncmp(args, "no", sizeof("no") - 1) ||
+             !strncmp(args, "false", sizeof("false") - 1)) {
+    return 0;
+  } else {
+    std::cerr << "Unrecognised argument for node-report verbose switch option: "
+              << args
+              << "\n";
+  }
+  return 0;  // Default is verbose mode off
+}
+
+/*******************************************************************************
+ * Function to save the node and subcomponent version strings. This is called
+ * during node-report module initialisation.
+ *******************************************************************************/
+void SetVersionString(Isolate* isolate) {
+  // Catch anything thrown and gracefully return
+  TryCatch trycatch(isolate);
+  version_and_command.version_string = NODE_VERSION_STRING;
+
+  // Retrieve the process object
+  Local<String> process_prop = String::NewFromUtf8(
+                                 Isolate::GetCurrent(),
+                                 "process",
+                                 NewStringType::kNormal)
+                                 .ToLocalChecked();
+  if (process_prop->IsNull()) return;
+
+  Local<Object> global_obj = isolate->GetCurrentContext()->Global();
+  EscapableHandleScope scope(isolate);
+  Local<Value> process_value = scope.Escape(
+                                         global_obj->Get(
+                                         isolate->GetCurrentContext(),
+                                         process_prop)
+                                         .FromMaybe(Local<Value>()));
+  if (process_value->IsNull()) return;
```

Comment:
It *seems* like this function should be replaced with something that doesn’t interact with JS at all, and instead use the same mechanism that Node.js uses to generate these values in the first place…

If that’s not feasible, then it seems like this should depend on the `Environment*` rather than looking up variables based on the current global scope of the `Isolate*`…

Reply:
+1. This function was necessary for the standalone npm module and had to be done up front (to avoid unnecessary calls into V8 during the report generation) but in core all of this information can be accessed directly. For example, this is the actual code used to populate the process object: 
https://github.com/nodejs/node/blob/f464ac3c74ca2b2411c578b4dcb52838f569e9fa/src/node.cc#L1868-L1932

and here's similar code for trace events:
https://github.com/nodejs/node/blob/f464ac3c74ca2b2411c578b4dcb52838f569e9fa/src/node.cc#L227-L266

(The duplicated openssl version slicing code is unfortunate -- Perhaps a separate PR could refactor it out.)

Reply:
sure, thanks for the pointers @addaleax @richardlau . will rework this part.

Reply:
this is done now, please have a look:
 - SetVersionString is replaced with GetComponentVersions that is directly consumed on demand.
 - For openssl, I am not bothering to cut out the numbers from the whole string, as we are forming a human-readable content. Hope this is fine by you.

Reply:
subsequent reviews and work resolved this fully.